core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -130,36 +117,23 @@ public AsyncNonBlockingStore(NonBlockingStore<K, V> actual) {
       segmentCount = storeConfiguration.segmented() ? cacheConfiguration.clustering().hash().numSegments() : 1;
       asyncConfiguration = storeConfiguration.async();
       modificationQueueSize = asyncConfiguration.modificationQueueSize();
-      // It is possible for multiple threads to write to this processor at the same time
-      submissionFlowable = UnicastProcessor.<Modification>create(1).toSerialized();
       nonBlockingExecutor = ctx.getNonBlockingExecutor();
-      startSub = submissionFlowable.window(requestFlowable).subscribe(this);
+      stopped = false;
       return actual.start(ctx);
    }
 
    @Override
    public CompletionStage<Void> stop() {
       CompletionStage<Void> asyncStage;
-      if (submissionFlowable != null) {
-         if (trace) {
-            log.tracef("Stopping async store containing store %s", actual);
-         }
-         submissionFlowable = null;
-         asyncStage = awaitQuiescence().whenComplete((ignore, t) -> {
-            // We can only dispose of the subscription after we are sure we are totally stopped
-            if (startSub != null) {
-               startSub.dispose();
-               startSub = null;
-            }
-         });
-
-      } else {
-         asyncStage = CompletableFutures.completedNull();
+      if (trace) {
+         log.tracef("Stopping async store containing store %s", actual);
       }
+      asyncStage = awaitQuiescence();
```

Comment:
Hrmm, I think it should be okay if I just move `stopped = true;` to the line before calling `awaitQuiescence`. There is still technically a tiny window, as this would really require a check before submitting to be 100%, but I am not too worried about that.

Reply:
I created https://issues.redhat.com/browse/ISPN-12647 to look more closely at the shutdown procedure in all the bundled store implementations.